shared-libs/cht-datasource/src/libs/core.ts

```diff
@@ -0,0 +1,44 @@
+import { DataContext } from './data-context';
+
+/**
+ * A value that could be `null`.
+ */
+export type Nullable<T> = T | null;
+
+/**
+ * An array that is guaranteed to have at least one element.
+ */
+export type NonEmptyArray<T> = [T, ...T[]];
+
+type DataPrimitive = string | number | boolean | Date | null | undefined;
+interface DataArray extends Readonly<(DataPrimitive | DataArray | DataObject)[]> { }
+
+/** @internal */
+export interface DataObject extends Readonly<Record<string, DataPrimitive | DataArray | DataObject>> { }
+
+/** @internal */
+export const isString = (value: unknown): value is string => {
+  return typeof value === 'string';
+};
+
+/** @internal */
+export const isRecord = (value: unknown): value is Record<string, unknown> => {
+  return value !== null && typeof value === 'object';
+};
+
+/** @internal */
+export const hasField = (value: Record<string, unknown>, field: { name: string, type: string }): boolean => {
+  const valueField = value[field.name];
+  return typeof valueField === field.type;
+};
+
+/** @internal */
+export const hasFields = (
+  value: Record<string, unknown>,
+  fields: NonEmptyArray<{ name: string, type: string }>
+): boolean => fields.every(field => hasField(value, field));
+
+/** @internal */
+export abstract class AbstractDataContext implements DataContext {
+  readonly get = <T>(fn: (ctx: DataContext) => T): T => fn(this);
+}
```

Comment:
I moved it to `shared-libs/cht-datasource/src/libs/data-context.ts` and changed the imports accordingly in the remote and local `data-context.ts` files and it doesn't seem to throw any error. Maybe something you did in recent commits fixed it?

Reply:
Hmm, I will have to try it again and see if I can recreate the error.  However, I did forget to mention that the error only happened _at runtime_ when running against the transpiled JS files.  So, `tsc` would run fine and my unit tests would work, but the api server would fail to start with an error about null constructors or something...

Reply:
Okay I eventually managed to reproduce this issue! Yes there *is* a circular dependency in `shared-libs/cht-datasource/src/libs/data-context.ts`, two actually:
- `import { isLocalDataContext } from '../local/libs/data-context';`
- `import { assertRemoteDataContext, isRemoteDataContext } from '../remote/libs/data-context';`

Those two files then import `DataContext` (and eventually `AbstractDataContext`) from `'../../libs/data-context'` which creates the circular dependency leaving all three exports from `src/libs/data-context.ts` undefined: `exports.adapt`, `exports.assertDataContext`, and `exports.AbstractDataContext`.

I think what you did is correct. The only other alternative is to move `isLocalDataContext`, `assertRemoteDataContext`, and `isRemoteDataContext` to `src/libs/data-context.ts`. Both way get us the same result so let's keep it the way it currently is :+1: